keps/sig-node/4622-topologymanager-max-allowable-numa-nodes/README.md

```diff
@@ -0,0 +1,708 @@
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+No.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+We add a metric: `kubelet_topology_manager_admission_time` for kubelet, which can be used to check if the setting is causing unacceptable performance drops.
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+Examine the kubelet configuration of a node to verify the existence of the feature gate and the utilization of the new policy option. we can use the following command to check the feature if it is enabled:
+
+```
+kubectl get --raw "/api/v1/nodes/<nodename>/proxy/configz" | jq '.kubeletconfig.TopologyManagerPolicyOptions'
+```
+
+###### How can someone using this feature know that it is working for their instance?
+
+<!--
+For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
+for each individual pod.
+Pick one more of these and delete the rest.
+Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
+and operation of this feature.
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+- [ ] Events
+  - Event Reason: 
+- [ ] API .status
+  - Condition name: 
+  - Other field: 
+- [x] Other (treat as last resort)
+  - Details: If their system has more than 8 NUMA nodes, the TopologyManager is turned on and the kubelet is not crashing, then the feature is working.
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+This feature is use-at-your-own-risk until we have a proper solution in place to reduce the state explosion. There is no good way to characterize the impact at the moment, as every system is different.
```

Comment:
This feature is going to beta and will become on-by-default for all clusters in 1.31?  If so, and I'm reading this right, I don't think this is an acceptable level of risk.  Have we considered remaining in alpha until a proper solution is in place?

Reply:
On by default just means "a user can opt-in to set a value higher than 8 for max-allowable-numa-nodes". It doesn't mean that systems with more than 8 NUMA nodes will suddenly have this feature enabled. A user will have to actively set max-allowable-numa-nodes to soe value greater than 8 for it to take action. The default action will still to crash the kubelet on such systems.

Reply:
I would be hesitant to do that because the whole point of this flag is to give people an "out" in situations where they are willing to accept a potential slowdown in order to not have their kubelet crash on them.

The limit of 8 was hard-coded previously, as a way to do exactly what you suggest, i.e. put a limit on the max value of the number of NUMA nodes that keeps the SLO below some expected time limit. This new flag is a way for people to ignore that and allow systems with a larger number of NUMA nodes to work if they are willing to accept the slowdown it (may) introduce.

Reply:
In other words -- the actual *value* of this new setting doesn't impact the latency of pod admission. The latency is *only* influenced by the number of NUMA nodes your physical system has. This flag exists solely to allow a user to increase `max-allowable-numa-nodes` to match the number of NUMA nodes you have on the system, so that the kubelet doesn't crash when bringing the TopologyManager online.

Reply:
@jpbetz I don't think @cyclinder will be back online before the KEP freeze later today. 

Can we agree to change the text above to the following text as a follow-up in the next day or two so that this PR can get merged:

> The value of `max-allowable-numa-nodes` does not (in and of itself) affect the latency of pod admission. With the TopologyManager enabled, the time to admit a pod is tied to the number of NUMA nodes on the physical machine. In the past, this was hard-coded at 8 to ensure that pod admission always completed in a reasonable amount of time. If a machine had more than 8 NUMA nodes, the kubelet would crash with a log message stating that the ToplogyManager is unsupported on machines with more than 8 NUMA nodes. With the new `max-allowable-numa-nodes` option, admins now have the ability to allow nodes with more than 8 NUMA nodes to run with the TopologyManager enabled. However, it is unknown exactly how much this will slow down pod admission on any given system. This feature is therefore to be used at-your-own-risk until we have a proper solution in place to reduce the state explosion that causes pod admission time to slow down as the number of NUMA nodes increases.

Reply:
I'm okay with agreeing on something to update the kep with.

The "used at-your-own-risk until we have a proper solution in place" wording seems to conflict with what a beta feature should be aiming to achieve.  Can we also recommend a soft limit (64 NUMA nodes or something that we have reason to believe is safe) and also agree to fix the "state explosion" problem before GA?

Reply:
Thanks @klueska @jpbetz!, updated.

> The "used at-your-own-risk until we have a proper solution in place" wording seems to conflict with what a beta feature should be aiming to achieve.

As @klueska explains: this OPTION solves the problem of kubelet not being able to start when the NUMA nodes are more than 8 nodes, and the risk is only that it might increase the time for the pod to start, I think the risk is low enough. since the benefits it brings are obvious enough.